src/libsyntax/codemap.rs

```diff
@@ -807,6 +807,83 @@ impl CodeMap {
                         hi.col.to_usize() + 1)).to_string()
     }
 
+    // Returns true if two spans have the same callee
+    // (Assumes the same ExpnFormat implies same callee)
+    fn match_callees(&self, sp_a: &Span, sp_b: &Span) -> bool {
+        let fmt_a = self
+            .with_expn_info(sp_a.expn_id,
+                            |ei| ei.map(|ei| ei.callee.format.clone()));
+
+        let fmt_b = self
+            .with_expn_info(sp_b.expn_id,
+                            |ei| ei.map(|ei| ei.callee.format.clone()));
+        fmt_a == fmt_b
+    }
+
+    /// Returns a formatted string showing the expansion chain of a span
+    ///
+    /// Spans are printed in the following format:
+    ///
+    /// filename:start_line:col: end_line:col
+    /// snippet
+    ///   CALLEE:
+    ///   Callee span, or 'None'
+    ///   CALLSITE:
+    ///   Callsite span, or 'None'
```

Comment:
Formatting of the output:
* I would not use caps for `callee`/`callsite` - no need to draw attention to the headings
* I would truncate the snippet to the smaller of 50 chars/one line + `...`
* rather than printing `None`, how about not printing anything at all (i.e., no heading or span)